### ISSUE TITLE
Convert C++14 [[deprecated]] attribute into target langauge

### ISSUE THREAD

**Issue by chrisdembia** [swig/swig#490]
With C++14, C++ now has a `[[deprecated]]` attribute to mark methods as deprecated. It'd be great if SWIG could pick up this attribute and turn it into something meaningful in the target language. For example, in python, we could issue a [`DeprecationWarning`](https://docs.python.org/2/library/exceptions.html#exceptions.DeprecationWarning).

**Comment by wsfulton** [comment]
Yes indeed. Parsing of the C++ attribute syntax would need to be implemented first and added to the parse tree for language modules like Python to pick up and use.

**Comment by dasvlmxclskfad** [comment]
is there any update on this?

**Comment by wsfulton** [comment]
Hi @dasvlmxclskfad. I'm not aware of anyone adding in C++11 attribute syntax to SWIG. How's your flex+bison knowledge? You'll need to be knowledgeable in these languages in order to modify the parser (parser.y) to parse the attributes. Once that is in, it would be quite easy to obtain this information from the Python module and convert into a Python `@deprecated`.

**Comment by dasvlmxclskfad** [comment]
I recently graduated and I have worked with them in the introductory compiler course that I took.

So I can read up more on those tools. Since you mentioned that there's is no support for C++11's attributes I would be great if you could point me the closest SWIG "feature" that I can refer to for implementation of the feature :)

**Comment by wsfulton** [comment]
Have a good read of the developer documentation, especially parse trees and the various `-debug` options at http://swig.org/Doc4.0/Extending.html and also the docs in the Docs/Devel directory.

You need to modify the parser to parse the attributes, wherever they are valid syntax and add to the parse tree. It's kind of similar to parsing Doxygen comments, except I suspect that Doxygen comments can occur in a wider number of places and note that parsing of Doxygen comments is off unless the `-doxygen` option is used. When completed, I'd expect the parse tree to show the optional attributes, so the output with the option`-debug-module 1` would have an extra attribute (called say cppattribs) amongst the other parse tree attributes, so you'd see the `cppattribs` in the output:
```
+++ cdecl - 0x7fe66fc67a90 ----------------------------------------
| sym:name     - "FooBar"
| name         - "FooBar"
| decl         - "f(int)."
| parms        - 'int x'
| kind         - "function"
| type         - "void"
| sym:symtab   - 0x7fe66fb36650
| sym:overname - "__SWIG_0"
| cppattribs   - "deprecated"
```
with input
```c++
[[deprecated]]
void FooBar(int x);
```
To begin with, simply adding the text between `[[` and `]]` into a string and adding to the parse tree will suffice. Decomposing the attributes into a function name and parameters could come later, so that more complex attributes can be handled, such as: `[[deprecated("Use BarFoo instead")]]`.

**Comment by ojwb** [comment]
Since e56ce01d SWIG skips any attributes in the scanner, so code using them can now be parsed by SWIG, but the attribute information is ignored entirely.

To implement this, the attributes would really need to be handled by the parser (as discussed above).

**Comment by ojwb** [comment]
Here's some notes about how to map this to some target languages:

* PHP: `trigger_error('World::hi() is deprecated, use World::hello() instead', E_USER_DEPRECATED);`
* Python: `import warnings;` ... `warnings.warn('World::hi() is deprecated, use World::hello() instead', DeprecationWarning)`
* Ruby: `warn '[DEPRECATION] World::hi() is deprecated, use World::hello() instead'` (simple option without using an external library to handle deprecation warnings)
* Perl: `use warnings;` ... `warnings::warnif('deprecated', 'World::hi() is deprecated, use World::hello() instead');`
* Java: `@Deprecated` in JavaDoc.

**Comment by vadz** [comment]
C#: `[Obsolete("use World::hello() instead")]`.

**Comment by medismailben** [comment]
Hey @ojwb! Any update on this ? Are there any blockers ? Thank you in advance :)

**Comment by ojwb** [comment]
I don't think there's really any update as such not noted above.

There was some discussion in a different ticket about doxygen comments which concluded that it would probably be best for the scanner to not generate tokens for doxygen comments and attributes but instead set global variables which the parser can then check to attach them to the appropriate node.  Doxygen comments can appear anywhere and we don't want them to cause a parse error if used in an odd place, with the result that parsing of them is off by default and you need to specify `-doxygen` which can then cause parse errors (e.g. #1715).

Attributes can't appear absolutely anywhere but can appear in many places and adjusting the parser grammar rules to fully handle them seems hard and liable to result in more conflicts in the grammar.  Doing it outside of the grammar probably actually makes implementing easier.

The blocker is really just lack of volunteer developer time.
